src/kubernetes.cc

```diff
@@ -1049,9 +1049,13 @@ void KubernetesReader::PodCallback(
     MetadataUpdater::UpdateCallback callback,
     const json::Object* pod, Timestamp collected_at, bool is_deleted) const
     throw(json::Exception) {
-  std::vector<MetadataUpdater::ResourceMetadata> result_vector =
-      GetPodAndContainerMetadata(pod, collected_at, is_deleted);
-  callback(std::move(result_vector));
+  try {
+    std::vector<MetadataUpdater::ResourceMetadata> result_vector =
+        GetPodAndContainerMetadata(pod, collected_at, is_deleted);
+    callback(std::move(result_vector));
+  } catch (const json::Exception& e) {
+    LOG(ERROR) << e.what();
+  }
```

Comment:
A couple of questions:

1. Why is the error not being caught + logged by the try-catch block around the Watchmaster (L1077)?
2. For parity, please add a try-catch block in the NodeCallback too

Reply:
+1. This should have been logged in `WatchMaster`. I think the issue here is that the thread dies, but the right fix is to restart the watch, not to swallow the error.